### PR TITLE
fix(51222): Go-to-definition on return statements should jump to the containing function declaration

### LINKED ISSUE #51222: Go-to-definition on `return` statements should jump to the containing function declaration
```ts
function foo() {
    return /*END*/() => {
        /*START*/return 10;
    }
}
```

Go to definition on `START` should jump you up to the containing function.

It should be resilient to questionable placements of `return` and do nothing, as in the following cases.

```ts
/*a*/return;


function foo() {
    class Foo {
       static { /*b*/return; }
    }
}
```

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #51222

### CODE REVIEW

File: tests/cases/fourslash/goToDefinitionReturn1.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+/// <reference path="fourslash.ts" />
+
+////function /*end*/foo() {
```
Comment (sheetalkamat) [inline tests/cases/fourslash/goToDefinitionReturn1.ts:3]: Can you please also add test with overloads. Thanks

**Review by sheetalkamat — APPROVED** [review]

File: src/services/goToDefinition.ts [hunk 2]
```diff
@@ -26,6 +26,12 @@ namespace ts.GoToDefinition {
             return label ? [createDefinitionInfoFromName(typeChecker, label, ScriptElementKind.label, node.text, /*containerName*/ undefined!)] : undefined; // TODO: GH#18217
         }
 
+        if (node.kind === SyntaxKind.ReturnKeyword) {
+            const functionDeclaration = findAncestor(node.parent, n =>
+                isClassStaticBlockDeclaration(n) ? "quit" : isFunctionLikeDeclaration(n)) as FunctionLikeDeclaration;
```
Comment (DanielRosenwasser) [inline src/services/goToDefinition.ts:31]: There is technically a `getContainingFunctionOrClassStaticBlock` - but I think this works too.

Comment (sheetalkamat) [reply src/services/goToDefinition.ts:31]: Sorry @DanielRosenwasser, I thought this was good to go so merged.

Comment (DanielRosenwasser) [reply src/services/goToDefinition.ts:31]: Nah, this is fine I think.

Comment (a-tarasyuk) [reply src/services/goToDefinition.ts:31]: `getContainingFunctionOrClassStaticBlock` uses `isFunctionLikeOrClassStaticBlockDeclaration -> isFunctionLikeKind`, which allows to use the following node types
```ts
case SyntaxKind.MethodSignature:
case SyntaxKind.CallSignature:
case SyntaxKind.JSDocSignature:
case SyntaxKind.ConstructSignature:
case SyntaxKind.IndexSignature:
case SyntaxKind.FunctionType:
case SyntaxKind.JSDocFunctionType:
case SyntaxKind.ConstructorType:
```

I think in this case we don't need all of them, or am I wrong?

Comment (a-tarasyuk) [reply src/services/goToDefinition.ts:31]: Okay :)

Comment (DanielRosenwasser) [inline src/services/goToDefinition.ts:31]: I guess one thing I wish was that this said

```suggestion
                isClassStaticBlockDeclaration(n) ? "quit" : isFunctionLikeDeclaration(n)) as FunctionLikeDeclaration | undefined;
```

**Pull request merged**